app/src/main/java/com/cdio/solitaire/controller/GameStateController.kt

```
@@ -0,0 +1,316 @@
+            GameState(foundations, tableaux, talon, stock, mutableListOf(Move(MoveType.DEAL_CARDS)))
+    }
+
+    /**
+     * Deals out cards in compliance with solitaire rules.
+     */
+    private fun dealOutDeck(
+        deck: CardStack,
+        tableaux: Array<CardStack>,
+        stock: CardStack
+    ) {
+        for (i in 0..6) {
+            for (j in i..6) {
+                tableaux[j].pushCard(deck.popCard()!!)
+                tableaux[j].hiddenCards++
+            }
+        }
+        stock.pushStack(deck)
+        stock.hiddenCards = stock.size
+    }
+
+    /**
+     * Used to create anonymous (hidden) cards.
+     */
+    private fun createNullCardStack(cardCount: Int, stackID: Int): CardStack {
+        val cardStack = CardStack(stackID)
+        for (i in 0..cardCount) cardStack.pushCard(Card(stackID))
+        return cardStack
+    }
+
+    /**
+     * Retrieve a cardStack using the ID. The numbering is in compliance with initial gameState.
+     */
+    fun getCardStackFromID(stackID: Int): CardStack? { // Added for ease, might be deleted later
+        return when (stackID) {
+            in 1..7 -> gameState.tableaux[stackID]
+            in 8..11 -> gameState.foundations[stackID - 7]
+            12 -> gameState.stock
+            0 -> gameState.talon
+            else -> null
+        }
+    }
+
+    /**
+     * Flips the talon by pushing the talon to the head end of stock.
+     */
+    private fun flipTalon() {
+        gameState.stock.pushStackToHead(gameState.talon)
+        gameState.stock.hiddenCards += gameState.talon.hiddenCards
+        gameState.talon.hiddenCards = 0
+    }
+
+    /**
+     * Draw 3 cards from stock. Do accounting on hiddenCards, and, if needed, signal card recognition.
+     */
+    // TODO: Consider legal move check.
+    private fun drawFromStock(): Card? {
+        if (gameState.stock.size < 3) {
+            Log.e("EmptyStackPop", "Not enough cards in stock to draw to talon.")
+            return null
```

Consider throwing an exception here instead of just logging the error, since a stack trace maybe could provide a lot of useful information.

That also means the method doesn't have to return a nullable object, which can make the code simpler.

---

The method can return a nullable object, because of cardToUpdate(). This is purposely chosen. :)